packages/rum-core/src/domain/rumEventsCollection/action/trackClickActions.ts

```diff
@@ -0,0 +1,229 @@
+          if (collectFrustrations) {
+            potentialClickAction.addFrustration(FrustrationType.DEAD)
+            potentialClickAction.validate()
+          } else {
+            potentialClickAction.discard()
+          }
+        } else if (idleEvent.end < startClocks.timeStamp) {
+          // If the clock is looking weird, just discard the action
+          potentialClickAction.discard()
+        } else {
+          // Else validate the potential click action at the end of the page activity
+          potentialClickAction.validate(idleEvent.end)
+        }
+        stopClickProcessing()
+      },
+      CLICK_ACTION_MAX_DURATION
+    )
+
+    let viewCreatedSubscription: Subscription | undefined
+    if (!collectFrustrations) {
+      // TODO: remove this in a future major version. To keep retrocompatibility, end the potential
+      // click action on a new view is created.
+      viewCreatedSubscription = lifeCycle.subscribe(LifeCycleEventType.VIEW_CREATED, stopClickProcessing)
+    }
+
+    const stopSubscription = stopObservable.subscribe(stopClickProcessing)
+
+    function stopClickProcessing() {
+      // Cleanup any ongoing process
+      potentialClickAction.discard()
+      if (viewCreatedSubscription) {
+        viewCreatedSubscription.unsubscribe()
+      }
+      stopWaitingIdlePage()
+      stopSubscription.unsubscribe()
+    }
+  }
+}
+
+function listenClickEvents(callback: (clickEvent: MouseEvent & { target: Element }) => void) {
+  return addEventListener(
+    window,
+    DOM_EVENT.CLICK,
+    (clickEvent: MouseEvent) => {
+      if (clickEvent.target instanceof Element) {
+        callback(clickEvent as MouseEvent & { target: Element })
+      }
+    },
+    { capture: true }
+  )
+}
+
+function newPotentialClickAction(
+  lifeCycle: LifeCycle,
+  history: ContextHistory<string>,
+  collectFrustrations: boolean,
+  base: Pick<ClickAction, 'startClocks' | 'event' | 'name'>
+) {
+  const id = generateUUID()
+  const historyEntry = history.add(id, base.startClocks.relative)
+  const eventCountsSubscription = trackEventCounts(lifeCycle)
+  let isStopped = false
+
+  const frustrations = new Set<FrustrationType>()
+
+  function stop(endTime?: TimeStamp) {
+    isStopped = true
+    if (endTime) {
+      historyEntry.close(getRelativeTime(endTime))
+    } else {
+      historyEntry.remove()
+    }
+    eventCountsSubscription.stop()
+  }
+
+  function addFrustration(frustration: FrustrationType) {
+    if (collectFrustrations) {
+      frustrations.add(frustration)
+    }
+  }
+
+  return {
+    addFrustration,
+
+    validate: (endTime?: TimeStamp) => {
+      if (isStopped) {
+        return
+      }
+      stop(endTime)
+      if (eventCountsSubscription.eventCounts.errorCount > 0) {
+        addFrustration(FrustrationType.ERROR)
+      }
+
+      const frustrationTypes: FrustrationType[] = []
+      frustrations.forEach((frustration) => {
+        frustrationTypes.push(frustration)
+      })
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏what about extracting a `toArray(set: Set)` utility method? it could remove a bit of plumbing from this function